packages/client/src/components/drawing-canvas.tsx

```diff
@@ -0,0 +1,171 @@
+        const c = n.content.getContent();
+        for (let i = 0; i < c.length; i++) {
+          cs.push(c[i]);
+        }
+      }
+    }
+    n = n.right;
+  }
+  return cs;
+};
+
+const getDeletedItems = (type, event) => {
+  const cs = [];
+  let n = type._start;
+  while (n !== null) {
+    if (event.deletes(n)) {
+      if (n.countable && n.deleted) {
+        const c = n.content.getContent();
+        for (let i = 0; i < c.length; i++) {
+          cs.push(c[i]);
+        }
+      }
+    }
+    n = n.right;
+  }
+  return cs;
+};
+
+const lines = new Map();
+
+let svg;
+
+function DrawingCanvas({ document_id, name }) {
+  const ref = React.useRef();
+  const ydoc = new Y.Doc();
+  const drawing = ydoc.getArray("drawing");
+  const renderPath = d3.svg
+    .line()
+    .x((d) => {
+      return d[0];
+    })
+    .y((d) => {
+      return d[1];
+    })
+    .interpolate("basis");
+
+  // create line from a shared array object and update the line when the array changes
+  const drawLine = (yarray) => {
+    const line = svg.append("path").datum(yarray.toArray());
+    line.attr("d", renderPath);
+    line.attr("fill", "transparent");
+    line.attr("style", "stroke:black;stroke-width:1");
+    lines.set(yarray, line);
+    yarray.observe((event) => {
+      const newItems = getInsertedItems(yarray, event);
+      // we only implement insert events that are appended to the end of the array
+      newItems.forEach((value) => {
+        line.datum().push(value);
+      });
+      line.attr("d", renderPath);
+    });
+  };
+
+  let sharedLine = null;
+  const dragStart = () => {
+    sharedLine = new Y.Array();
+    drawing.insert(drawing.length, [sharedLine]);
+  };
+
+  // After one dragged event is recognized, we ignore them for 33ms.
+  let ignoreDrag = null;
+  function drag() {
+    if (sharedLine != null && ignoreDrag == null) {
+      ignoreDrag = setTimeout(() => {
+        ignoreDrag = null;
+      }, 33);
+      const coords = d3.mouse(this);
+      sharedLine.push([coords]);
+    }
+  }
+
+  const dragEnd = () => {
+    sharedLine = null;
+    clearTimeout(ignoreDrag);
+    ignoreDrag = null;
+  };
+
+  const undoManager = () => {
+    const item = drawing.get(drawing.length - 1)["_item"];
+    if (drawing.length > 0 && item.id.client === ydoc.clientID) {
+      drawing.delete(drawing.length - 1, 1);
+    }
+  };
+
+  useEffect(() => {
+    svg = d3.select("#drawingCanvas");
+    const provider = new WebrtcProvider(document_id, ydoc, {
+      signaling: [SIGNALLING_SERVICE],
+    } as any);
+    const wsProvider = new WebsocketProvider(
+      PROVIDER_SERVICE,
+      document_id,
+      ydoc
+    );
+    console.log("BSDFVASDF");
+    wsProvider.on("status", (event) => {
+      console.log(event.status); // logs "connected" or "disconnected"
+    });
+    provider.awareness.setLocalStateField("user", {
+      color: `#${Math.floor(Math.random() * 16777215).toString(16)}`,
+      name,
+    });
```

Comment:
A lot of this stuff is getting repeated from the editor. Probably a good idea to extract it. Maybe keep it in its own `useYDoc(id)` hook?